_posts/2021-11-30-pravega-connector-101.md

```diff
@@ -0,0 +1,213 @@
+
+Here comes Pravega that supports [transactional writes](https://cncf.pravega.io/docs/latest/transactions/). The idea with a Pravega transaction is that it allows an application to prepare a set of events that can be written "all at once" to a Stream. This allows an application to "commit" a bunch of events atomically. With this Pravega feature making the writes idempotent, it is possible to implement end-to-end exactly-once pipelines together with Flink. 
+
+To build such an end-to-end solution is still not easy, the main difficulty is to have the coordination between Flink and Pravega sink. One common approach for coordinating commits and rollbacks in a distributed system is the two-phase commit protocol. We followed this path, worked together with the Flink community and implemented the sink function in a two-phase commit way coordinated with Flink checkpoints.
+
+Flink community then extracted the common logic of the two-phase commit protocol and provided a general interface `TwoPhaseCommitSinkFunction` (relevant Jira [here](https://issues.apache.org/jira/browse/FLINK-7210)) to make it possible to build end-to-end exactly-once applications with other message system with transaction support including Apache Kafka versions 0.11 and beyond. There is an official Flink [blog](https://flink.apache.org/features/2018/03/01/end-to-end-exactly-once-apache-flink.html) as well to introduce this feature in detail.
+
+# Wrap Up and plan for the future
+
+If you’ve made it this far, thanks for staying with us through a detailed post. Pravega Flink connector is a project that makes Pravega connects to Flink and breathe life into Pravega to act as a key data store in a streaming pipeline. Here are some key points that we covered
+
+- The background and history of Flink and the connector integration
+- Introduction to basic usage of the streaming source and sink in Pravega Flink connector
+- Deep dive into the internals of the checkpoint and end-to-end exactly-once feature
+
+Finally I'd like to mention some of our plans for the future work on Flink connector.
+
+- `FlinkPravegaInputFormat` and `FlinkPravegaOutputFormat` is now provided to support batch reads and writes in Flink, but these are under legacy DataSet API. As Flink is now making much effort to unify batch and streaming, it has dedicated on improving from API level recently and provided new interfaces for [source](https://cwiki.apache.org/confluence/display/FLINK/FLIP-27%3A+Refactor+Source+Interface) and [sink](https://cwiki.apache.org/confluence/display/FLINK/FLIP-143%3A+Unified+Sink+API) API in the recent Flink 1.11 and 1.12 releases. We will still actively communicate with Flink community and integrate with the new APIs. 
+- We will also put more effort on SQL / Table API support to provide a better using experience as it's simpler to understand and even more powerful to use in some cases.
```

Comment:
```suggestion
# Summary

The Pravega Flink connector enables Pravega to connect to Flink and allows Pravega to act as a key data store in a streaming pipeline. Both projects share a common design philosophy and can integrate well with each other. Pravega has its own concept of checkpointing and has implemented transactional writes to support end-to-end exactly-once guarantees. 

# Future plans

`FlinkPravegaInputFormat` and `FlinkPravegaOutputFormat` are now provided to support batch reads and writes in Flink, but these are under the legacy DataSet API. Since Flink is now making efforts to unify batch and streaming, it is improving its APIs and providing new interfaces for the [source](https://cwiki.apache.org/confluence/display/FLINK/FLIP-27%3A+Refactor+Source+Interface) and [sink](https://cwiki.apache.org/confluence/display/FLINK/FLIP-143%3A+Unified+Sink+API) APIs in the Flink 1.11 and 1.12 releases. We will continue to work with the Flink community and integrate with the new APIs.
 
We will also put more effort into SQL / Table API support in order to provide a better user experience since it is simpler to understand and even more powerful to use in some cases.
```